compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertCoreForallToFor.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file the conversion of `scf.forall` into `scf.for` with `aie.core`
+// operations. This is needed for now for further lowering as `scf.forall` isn't
+// supported in the further lowering of `aie.core`.
+//
+//===----------------------------------------------------------------------===//
+
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+
+#define DEBUG_TYPE "iree-amdaie-convert-core-forall-to-for"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+LogicalResult coreForallToFor(RewriterBase &rewriter,
+                              xilinx::AIE::CoreOp coreOp) {
+  WalkResult res = coreOp->walk([&](scf::ForallOp forallOp) {
+    SmallVector<Operation *> forOpResults;
+    if (failed(scf::forallToForLoop(rewriter, forallOp, &forOpResults))) {
+      coreOp.emitOpError() << "failed to transform scf.forall to scf.for";
+      return WalkResult::interrupt();
+    }
+
+    SmallVector<scf::ForOp> forOps = llvm::map_to_vector(
+        forOpResults,
+        [](Operation *loop) { return dyn_cast<scf::ForOp>(loop); });
+    if (failed(coalesceLoops(rewriter, forOps))) {
```

Comment:
Please add a test with 1) perfectly nested loops and 2) not perfectly nested loops, and add comment in pass description that you also coalesce. Or don't coalesce :)